File: kotlin/src/main/kotlin/io/github/ocirne/aoc/year2024/Day13.kt
package io.github.ocirne.aoc.year2024

import io.github.ocirne.aoc.AocChallenge

class Day13(val lines: List<String>) : AocChallenge(2024, 13) {

    private val buttonA = Regex("Button A: X\\+(\\d+), Y\\+(\\d+)")
    private val buttonB = Regex("Button B: X\\+(\\d+), Y\\+(\\d+)")
    private val prize = Regex("Prize: X=(\\d+), Y=(\\d+)")

    private fun calculateCost(ax: Long, ay: Long, bx: Long, by: Long, px: Long, py: Long): Long {
        val t1 = ay * px - ax * py
        val t2 = ay * bx - ax * by
        if (t1 % t2 != 0L)
            return 0
        val b = t1 / t2
        val t3 = px - b * bx
        if (t3 % ax != 0L)
            return 0
        val a = t3 / ax
        if (a < 0 || b < 0)
            return 0
        return b + 3 * a
    }

    private fun pressButtons(offset: Long = 0L): Long {
        return lines.chunked(4).sumOf { (a, b, p, _) ->
            val (ax, ay) = buttonA.find(a)!!.destructured
            val (bx, by) = buttonB.find(b)!!.destructured
            val (px, py) = prize.find(p)!!.destructured
            calculateCost(
                ax.toLong(), ay.toLong(),
                bx.toLong(), by.toLong(),
                px.toLong() + offset, py.toLong() + offset
            )
        }
    }

    override fun part1(): Long {
        return pressButtons()
    }

    override fun part2(): Long {
        return pressButtons(10000000000000L)
    }
}

File: python/src/aoc/year2024/day13.py
from aoc.util import load_input, load_example

import re

BUTTON_A_PATTERN = r"Button A: X\+(\d+), Y\+(\d+)"
BUTTON_B_PATTERN = r"Button B: X\+(\d+), Y\+(\d+)"
PRIZE_PATTERN = r"Prize: X=(\d+), Y=(\d+)"


def calculate_cost(ax, ay, bx, by, px, py):
    b = (ay * px - ax * py) / (ay * bx - ax * by)
    a = (px - b * bx) / ax
    return int(b + 3 * a) if 0 < a == int(a) and 0 < b == int(b) else 0


def press_buttons(lines, offset=0):
    total = 0
    for line in lines:
        a = re.search(BUTTON_A_PATTERN, line)
        if a:
            ax, ay = map(int, a.groups())
        b = re.search(BUTTON_B_PATTERN, line)
        if b:
            bx, by = map(int, b.groups())
        p = re.search(PRIZE_PATTERN, line)
        if p:
            px, py = map(int, p.groups())
            total += calculate_cost(ax, ay, bx, by, px + offset, py + offset)
    return total


def part1(lines):
    """
    >>> part1(load_example(__file__, "13"))
    480
    """
    return press_buttons(lines)


def part2(lines):
    return press_buttons(lines, 10000000000000)


if __name__ == "__main__":
    data = load_input(__file__, 2024, "13")
    print(part1(data))
    print(part2(data))
